diff --git a/tests/runtests.py b/tests/runtests.py
index f1a194d..0b27dd4 100755
--- a/tests/runtests.py
+++ b/tests/runtests.py
@@ -20,6 +20,24 @@ except ImportError as e:
         "Django module not found, reference tests/README.rst for instructions."
     ) from e
 else:
+    # Monkeypatch Django template engine with Rusty Templates
+    import django_rusty_templates
+
+    import django.template
+    import django.template.backends.django
+    import django.template.engine
+
+    # Replace Engine class but keep the `get_default` staticmethod
+    # that will return the correct `Engine`
+    django_get_default = django.template.Engine.get_default
+    rust_engine = django_rusty_templates.Engine
+    rust_engine.get_default = django_get_default
+    django.template.Engine = django.template.engine.Engine = rust_engine
+
+    django.template.backends.django.DjangoTemplates = (
+        django_rusty_templates.RustyTemplates
+    )
+
     from django.apps import apps
     from django.conf import settings
     from django.core.exceptions import ImproperlyConfigured
diff --git a/tests/template_loader/tests.py b/tests/template_loader/tests.py
index 569f86b..f482fb7 100644
--- a/tests/template_loader/tests.py
+++ b/tests/template_loader/tests.py
@@ -44,7 +44,6 @@ class TemplateLoaderTests(SimpleTestCase):
             e.exception.chain[-1].tried[0][0].template_name,
             "template_loader/unknown.html",
         )
-        self.assertEqual(e.exception.chain[-1].backend.name, "django")
 
     def test_select_template_first_engine(self):
         template = select_template(
@@ -92,7 +91,6 @@ class TemplateLoaderTests(SimpleTestCase):
             e.exception.chain[-1].tried[0][0].template_name,
             "template_loader/missing.html",
         )
-        self.assertEqual(e.exception.chain[-1].backend.name, "django")
 
     def test_select_template_tries_all_engines_before_names(self):
         template = select_template(
@@ -124,7 +122,6 @@ class TemplateLoaderTests(SimpleTestCase):
             e.exception.chain[-1].tried[0][0].template_name,
             "template_loader/unknown.html",
         )
-        self.assertEqual(e.exception.chain[-1].backend.name, "django")
 
     def test_render_to_string_with_list_first_engine(self):
         content = render_to_string(
@@ -163,7 +160,6 @@ class TemplateLoaderTests(SimpleTestCase):
             e.exception.chain[1].tried[0][0].template_name,
             "template_loader/unknown.html",
         )
-        self.assertEqual(e.exception.chain[1].backend.name, "django")
         self.assertEqual(
             e.exception.chain[2].tried[0][0].template_name,
             "template_loader/missing.html",
@@ -173,7 +169,6 @@ class TemplateLoaderTests(SimpleTestCase):
             e.exception.chain[3].tried[0][0].template_name,
             "template_loader/missing.html",
         )
-        self.assertEqual(e.exception.chain[3].backend.name, "django")
 
     def test_render_to_string_with_list_tries_all_engines_before_names(self):
         content = render_to_string(
diff --git a/tests/template_tests/syntax_tests/test_basic.py b/tests/template_tests/syntax_tests/test_basic.py
index 50e7a4c..75aef7e 100644
--- a/tests/template_tests/syntax_tests/test_basic.py
+++ b/tests/template_tests/syntax_tests/test_basic.py
@@ -327,7 +327,7 @@ class BasicSyntaxTests(SimpleTestCase):
 
     @setup({"template": "{% block content %}"})
     def test_unclosed_block(self):
-        msg = "Unclosed tag on line 1: 'block'. Looking for one of: endblock."
+        msg = "Unclosed 'block' tag. Looking for one of: 'endblock', 'endblock content'"
         with self.assertRaisesMessage(TemplateSyntaxError, msg):
             self.engine.render_to_string("template")
 
diff --git a/tests/template_tests/syntax_tests/test_extends.py b/tests/template_tests/syntax_tests/test_extends.py
index 9f5dae0..326aeb9 100644
--- a/tests/template_tests/syntax_tests/test_extends.py
+++ b/tests/template_tests/syntax_tests/test_extends.py
@@ -437,7 +437,7 @@ class InheritanceTests(SimpleTestCase):
     @setup(inheritance_templates)
     def test_inheritance_empty(self):
         with self.assertRaisesMessage(
-            TemplateSyntaxError, "'extends' takes one argument"
+            TemplateSyntaxError, "Expected an argument"
         ):
             self.engine.render_to_string("inheritance_empty")
 
diff --git a/tests/template_tests/syntax_tests/test_include.py b/tests/template_tests/syntax_tests/test_include.py
index be0deee..4d6657b 100644
--- a/tests/template_tests/syntax_tests/test_include.py
+++ b/tests/template_tests/syntax_tests/test_include.py
@@ -42,7 +42,7 @@ class IncludeTagTests(SimpleTestCase):
     def test_include04(self):
         template = self.engine.get_template("include04")
         with self.assertRaises(TemplateDoesNotExist):
-            template.render(Context({}))
+            template.render({})
 
     @setup(
         {
@@ -197,34 +197,31 @@ class IncludeTagTests(SimpleTestCase):
     def test_include_error07(self):
         template = self.engine.get_template("include-error07")
         with self.assertRaises(RuntimeError):
-            template.render(Context())
+            template.render({})
 
     @setup({"include-error08": '{% include "include-fail2" %}'}, include_fail_templates)
     def test_include_error08(self):
         template = self.engine.get_template("include-error08")
         with self.assertRaises(TemplateSyntaxError):
-            template.render(Context())
+            template.render({})
 
     @setup({"include-error09": "{% include failed_include %}"}, include_fail_templates)
     def test_include_error09(self):
-        context = Context({"failed_include": "include-fail1"})
+        context = {"failed_include": "include-fail1"}
         template = self.engine.get_template("include-error09")
         with self.assertRaises(RuntimeError):
             template.render(context)
 
     @setup({"include-error10": "{% include failed_include %}"}, include_fail_templates)
     def test_include_error10(self):
-        context = Context({"failed_include": "include-fail2"})
+        context = {"failed_include": "include-fail2"}
         template = self.engine.get_template("include-error10")
         with self.assertRaises(TemplateSyntaxError):
             template.render(context)
 
     @setup({"include_empty": "{% include %}"})
     def test_include_empty(self):
-        msg = (
-            "'include' tag takes at least one argument: the name of the "
-            "template to be included."
-        )
+        msg = "Expected an argument"
         with self.assertRaisesMessage(TemplateSyntaxError, msg):
             self.engine.get_template("include_empty")
 
@@ -238,7 +235,7 @@ class IncludeTests(SimpleTestCase):
         engine = Engine(app_dirs=True, debug=True)
         template = engine.get_template("test_include_error.html")
         with self.assertRaisesMessage(TemplateDoesNotExist, "missing.html"):
-            template.render(Context())
+            template.render({})
 
     def test_extends_include_missing_baseloader(self):
         """
@@ -249,7 +246,7 @@ class IncludeTests(SimpleTestCase):
         engine = Engine(app_dirs=True, debug=True)
         template = engine.get_template("test_extends_error.html")
         with self.assertRaisesMessage(TemplateDoesNotExist, "missing.html"):
-            template.render(Context())
+            template.render({})
 
     def test_extends_include_missing_cachedloader(self):
         engine = Engine(
@@ -266,23 +263,21 @@ class IncludeTests(SimpleTestCase):
 
         template = engine.get_template("test_extends_error.html")
         with self.assertRaisesMessage(TemplateDoesNotExist, "missing.html"):
-            template.render(Context())
+            template.render({})
 
         # Repeat to ensure it still works when loading from the cache
         template = engine.get_template("test_extends_error.html")
         with self.assertRaisesMessage(TemplateDoesNotExist, "missing.html"):
-            template.render(Context())
+            template.render({})
 
     def test_include_template_argument(self):
         """
         Support any render() supporting object
         """
         engine = Engine()
-        ctx = Context(
-            {
-                "tmpl": engine.from_string("This worked!"),
-            }
-        )
+        ctx = {
+            "tmpl": engine.from_string("This worked!"),
+        }
         outer_tmpl = engine.from_string("{% include tmpl %}")
         output = outer_tmpl.render(ctx)
         self.assertEqual(output, "This worked!")
@@ -296,15 +291,15 @@ class IncludeTests(SimpleTestCase):
         ]
         for template_names in tests:
             with self.subTest(template_names):
-                output = outer_temp.render(Context({"var": template_names}))
+                output = outer_temp.render({"var": template_names})
                 self.assertEqual(output, "index\n")
 
     def test_include_template_none(self):
         engine = Engine.get_default()
         outer_temp = engine.from_string("{% include var %}")
-        ctx = Context({"var": None})
-        msg = "No template names provided"
-        with self.assertRaisesMessage(TemplateDoesNotExist, msg):
+        ctx = {"var": None}
+        msg = "invalid template name: None"
+        with self.assertRaisesMessage(TypeError, msg):
             outer_temp.render(ctx)
 
     def test_include_from_loader_get_template(self):
@@ -335,7 +330,7 @@ class IncludeTests(SimpleTestCase):
             t = engine.get_template("recursive_include.html")
             self.assertEqual(
                 "Recursion!  A1  Recursion!  B1   B2   B3  Recursion!  C1",
-                t.render(Context({"comments": comments}))
+                t.render({"comments": comments})
                 .replace(" ", "")
                 .replace("\n", " ")
                 .strip(),
@@ -345,7 +340,7 @@ class IncludeTests(SimpleTestCase):
             t = engine.get_template("recursive_relative_include.html")
             self.assertEqual(
                 "Recursion!  A1  Recursion!  B1   B2   B3  Recursion!  C1",
-                t.render(Context({"comments": comments}))
+                t.render({"comments": comments})
                 .replace(" ", "")
                 .replace("\n", " ")
                 .strip(),
@@ -361,7 +356,7 @@ class IncludeTests(SimpleTestCase):
             """
             outer_tmpl = engine.from_string("{% include tmpl %}")
             output = outer_tmpl.render(
-                Context({"tmpl": engine.from_string(template), "comments": comments})
+                {"tmpl": engine.from_string(template), "comments": comments}
             )
             self.assertEqual(
                 "Recursion!  A1  Recursion!  B1   B2   B3  Recursion!  C1",
diff --git a/tests/template_tests/test_extends.py b/tests/template_tests/test_extends.py
index 0d2a934..7688176 100644
--- a/tests/template_tests/test_extends.py
+++ b/tests/template_tests/test_extends.py
@@ -1,6 +1,6 @@
 import os
 
-from django.template import Context, Engine, TemplateDoesNotExist, TemplateSyntaxError
+from django.template import Engine, TemplateDoesNotExist, TemplateSyntaxError
 from django.test import SimpleTestCase
 
 from .utils import ROOT, setup
@@ -12,7 +12,7 @@ class ExtendsBehaviorTests(SimpleTestCase):
     def test_normal_extend(self):
         engine = Engine(dirs=[os.path.join(RECURSIVE, "fs")])
         template = engine.get_template("one.html")
-        output = template.render(Context({}))
+        output = template.render({})
         self.assertEqual(output.strip(), "three two one")
 
     def test_extend_recursive(self):
@@ -24,18 +24,18 @@ class ExtendsBehaviorTests(SimpleTestCase):
             ]
         )
         template = engine.get_template("recursive.html")
-        output = template.render(Context({}))
+        output = template.render({})
         self.assertEqual(output.strip(), "fs3/recursive fs2/recursive fs/recursive")
 
     def test_extend_missing(self):
         engine = Engine(dirs=[os.path.join(RECURSIVE, "fs")])
         template = engine.get_template("extend-missing.html")
         with self.assertRaises(TemplateDoesNotExist) as e:
-            template.render(Context({}))
+            template.render({})
 
         tried = e.exception.tried
         self.assertEqual(len(tried), 1)
-        self.assertEqual(tried[0][0].template_name, "missing.html")
+        self.assertTrue(tried[0][0].endswith("missing.html"))
 
     def test_recursive_multiple_loaders(self):
         engine = Engine(
@@ -62,7 +62,7 @@ class ExtendsBehaviorTests(SimpleTestCase):
             ],
         )
         template = engine.get_template("one.html")
-        output = template.render(Context({}))
+        output = template.render({})
         self.assertEqual(
             output.strip(), "three locmem-three two locmem-two one locmem-one"
         )
@@ -75,11 +75,11 @@ class ExtendsBehaviorTests(SimpleTestCase):
         engine = Engine(dirs=[os.path.join(RECURSIVE, "fs")])
         template = engine.get_template("self.html")
         with self.assertRaises(TemplateDoesNotExist) as e:
-            template.render(Context({}))
+            template.render({})
         tried = e.exception.tried
         self.assertEqual(len(tried), 1)
         origin, message = tried[0]
-        self.assertEqual(origin.template_name, "self.html")
+        self.assertTrue(origin.endswith("self.html"))
         self.assertEqual(message, "Skipped to avoid recursion")
 
     def test_extend_cached(self):
@@ -99,7 +99,7 @@ class ExtendsBehaviorTests(SimpleTestCase):
             ],
         )
         template = engine.get_template("recursive.html")
-        output = template.render(Context({}))
+        output = template.render({})
         self.assertEqual(output.strip(), "fs3/recursive fs2/recursive fs/recursive")
 
         cache = engine.template_loaders[0].get_template_cache
@@ -109,7 +109,7 @@ class ExtendsBehaviorTests(SimpleTestCase):
 
         # Render another path that uses the same templates from the cache
         template = engine.get_template("other-recursive.html")
-        output = template.render(Context({}))
+        output = template.render({})
         self.assertEqual(output.strip(), "fs3/recursive fs2/recursive fs/recursive")
 
         # Template objects should not be duplicated.
@@ -144,7 +144,7 @@ class ExtendsBehaviorTests(SimpleTestCase):
             ]
         )
         template = engine.get_template("base.html")
-        output = template.render(Context({}))
+        output = template.render({})
         self.assertEqual(output.strip(), "loader2 loader1")
 
     def test_block_override_in_extended_included_template(self):
@@ -180,7 +180,7 @@ class ExtendsBehaviorTests(SimpleTestCase):
             ],
         )
         template = engine.get_template("base.html")
-        self.assertEqual(template.render(Context({})), "12AB")
+        self.assertEqual(template.render({}), "12AB")
 
     @setup(
         {"index.html": "{% block content %}B{% endblock %}{% extends 'base.html' %}"}
diff --git a/tests/template_tests/test_extends_relative.py b/tests/template_tests/test_extends_relative.py
index b20b015..b501df3 100644
--- a/tests/template_tests/test_extends_relative.py
+++ b/tests/template_tests/test_extends_relative.py
@@ -54,7 +54,7 @@ class ExtendsRelativeBehaviorTests(SimpleTestCase):
     def test_extend_error(self):
         engine = Engine(dirs=[RELATIVE])
         msg = (
-            "The relative path '\"./../two.html\"' points outside the file "
+            "The relative path './../two.html' points outside the file "
             "hierarchy that template 'error_extends.html' is in."
         )
         with self.assertRaisesMessage(TemplateSyntaxError, msg):
@@ -83,7 +83,7 @@ class IncludeRelativeBehaviorTests(SimpleTestCase):
     def test_include_error(self):
         engine = Engine(dirs=[RELATIVE])
         msg = (
-            "The relative path '\"./../three.html\"' points outside the file "
+            "The relative path './../three.html' points outside the file "
             "hierarchy that template 'error_include.html' is in."
         )
         with self.assertRaisesMessage(TemplateSyntaxError, msg):
@@ -105,10 +105,5 @@ class ExtendsMixedBehaviorTests(SimpleTestCase):
 
     def test_mixing_loop(self):
         engine = Engine(dirs=[RELATIVE])
-        msg = (
-            "The relative path '\"./dir2/../looped.html\"' was translated to "
-            "template name 'dir1/looped.html', the same template in which "
-            "the tag appears."
-        )
-        with self.assertRaisesMessage(TemplateSyntaxError, msg):
+        with self.assertRaises(TemplateSyntaxError):
             engine.render_to_string("dir1/looped.html")
diff --git a/tests/template_tests/tests.py b/tests/template_tests/tests.py
index 7364c7c..38fa883 100644
--- a/tests/template_tests/tests.py
+++ b/tests/template_tests/tests.py
@@ -70,10 +70,7 @@ class TemplateTestMixin:
         English.
         """
         engine = self._engine()
-        msg = (
-            "Invalid block tag on line 1: 'endblock', expected 'elif', 'else' "
-            "or 'endif'. Did you forget to register or load this tag?"
-        )
+        msg = "Unexpected tag 'endblock', expected elif, else or endif"
         with self.settings(USE_I18N=True), translation.override("de"):
             with self.assertRaisesMessage(TemplateSyntaxError, msg):
                 engine.from_string("{% if 1 %}lala{% endblock %}{% endif %}")
